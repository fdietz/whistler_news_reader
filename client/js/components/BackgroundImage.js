import React, { PropTypes, Component } from 'react';
import classNames from 'classnames';

class BackgroundImage extends Component {

  static propTypes = {
    imageUrl: PropTypes.string.isRequired,
    className: PropTypes.string,
    style: PropTypes.object,
  }

  constructor(props) {
    super(props);

    this.state = { loaded: false, error: false };

    this.onImageLoad = this.onImageLoad.bind(this);
  }

  componentDidMount() {
    const { imageUrl } = this.props;
    const img = new window.Image();
    img.onload = () => this.onImageLoad();
    img.onerror = () => this.setState({ error: true });
    img.src = imageUrl;
  }

  onImageLoad() {
    const imgTag = this.imgRef;
    const { imageUrl } = this.props;
    imgTag.style.backgroundImage = `url(${imageUrl})`;

    this.setState({ loaded: true });
  }

  render() {
    const { className, style } = this.props;

    const cls = classNames('background-image', className, {
      'image-loaded': this.state.loaded,
      'image-load-error': this.state.error
    });

    return (
      <div ref={(r) => { this.imgRef = r; }} className={cls} style={style} />
    );
  }
}

export default BackgroundImage;
