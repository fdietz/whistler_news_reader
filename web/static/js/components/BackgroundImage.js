import React, { PropTypes, Component } from "react";
import ReactDOM from "react-dom";
import classNames from "classnames";

class BackgroundImage extends Component {

  static propTypes = {
    imageUrl: PropTypes.string.isRequired,
    className: PropTypes.string,
    style: PropTypes.object
  }

  constructor(props) {
    super(props);

    this.state = { loaded: false };

    this.onImageLoad = this.onImageLoad.bind(this);
  }

  componentDidMount() {
    const { imageUrl } = this.props;
    const img = new window.Image();
    img.onload = () => this.onImageLoad();
    img.src = imageUrl;
  }

  onImageLoad() {
    const imgTag = ReactDOM.findDOMNode(this.refs.img);
    const { imageUrl } = this.props;
    imgTag.style.backgroundImage = `url(${imageUrl})`;

    this.setState({ loaded: true });
  }

  render() {
    const { className, style, ...rest } = this.props;

    const cls = classNames("background-image", className, {
      "image-loaded": this.state.loaded
    });

    return (
      <div ref="img" className={cls} style={style} {...rest}/>
    );
  }
}

export default BackgroundImage;
