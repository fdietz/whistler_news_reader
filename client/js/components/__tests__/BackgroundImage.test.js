import renderer from 'react-test-renderer';
import React from 'react';
import BackgroundImage from '../BackgroundImage';

describe('BackgroundImage', () => {
  it('renders correctly', () => {
    const tree = renderer.create(
      <BackgroundImage imageUrl="hello" />
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
