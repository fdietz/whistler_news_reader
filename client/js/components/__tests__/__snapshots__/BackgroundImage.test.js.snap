exports[`BackgroundImage renders correctly 1`] = `
<div
  className="background-image"
  style={undefined} />
`;
